### PR TITLE
Properly set source location of bare blocks.

### LINKED ISSUE #11440: Source map jumps back to function definition after bare block
## Description

This is an issue with source mapping.  Suppose you have a block inside of a function, and this block declares some local variables.  After the block is finished, these variables need to be popped off the stack.  But the way the resulting `POP` instructions get mapped is a little messed-up.

If the block is an `if`, `while`, or `do` block, then these `POP`s get mapped to `if`, `while` or `do` statement that the block is a part of.  This isn't a problem.  However, if the block is simply a bare block, not part of a larger control structure, these `POP`s will be mapped to the function definition node.  This is confusing.

I would suggest that as an alternative, these `POP`s could be mapped to the block node itself.

(Context: I noticed this behavior while investigating [this Truffle issue](https://github.com/trufflesuite/truffle/issues/4055) that resulted from it.  I've fixed the problem with Truffle Debugger, so this should no longer cause problems there, but sourcemapping things this way still seems like a bug so I thought I'd file an issue for it.)

## Environment

- Compiler version: 0.8.4 (but exists as far back as at least 0.5.0...)
- Target EVM version (as per compiler settings): default
- Framework/IDE (e.g. Truffle or Remix): Truffle

## Steps to Reproduce

I'll use code of the sort that was used in reporting the original Truffle issue:

```solidity
//SPDX-License-Identifier: MIT
pragma solidity ^0.8.4;
contract Example {
  event Something(uint);
  function run(uint x) public {
    {
      uint something = 3;
      emit Something(something);
    }
    emit Something(x);
  }
}
```

### PR DESCRIPTION
**Pull request by chriseth**
Fixes https://github.com/ethereum/solidity/issues/11440

### CODE REVIEW

**Comment by chriseth** [comment]
@haltman-at @yann300 This is a fix for #11440 but might have other consequences.

We never set the source location of a block itself, only of the statements inside the block. The only code that this could be relevant is the POPs for cleaning up local variables in the stack. This might also create source locations for the body of a function, though, which means that some POPs at the end of a function do not point to the function definition anymore but to the function body. This could also have similar consequences for any statement that contains a block like an if statement, while loop, for loop etc.

It would be great if you could check if everything still works for you.

**Review by leonardoalt — COMMENTED** [review]
Looks good, but will wait for tooling confirmation before approving.

**Comment by leonardoalt** [comment]
Tests need update

**Comment by haltman-at** [comment]
I haven't tested this but don't expect it to cause any problems for us?  If you really want me to test it the easiest way would be if there's a soljson.  (I can try building solc from source later but I ran into some problems with that last time I tried it.)  Still, I'm not sure how much testing this really needs wrt Truffle... Truffle Debugger should be perfectly fine with `POP`s at the end of a function pointing to the body rather than the definition.

**Comment by leonardoalt** [comment]
@haltman-at sounds good, here's a link to this PR's soljson: https://728857-40892817-gh.circle-artifacts.com/0/soljson.js

**Comment by haltman-at** [comment]
OK, tested it to be sure, all tests passing.  (OK, one test failing, but that's due to an unrelated matter. :P )  No problems as expected.  Didn't try making any new tests but I think our existing tests probably would have caught it were there a problem.

**Comment by yann300** [comment]
This build is also working ok in our side, we do have some issue but not related.

File: libsolidity/codegen/ContractCompiler.cpp [hunk 1]
```diff
@@ -1362,6 +1363,7 @@ bool ContractCompiler::visit(Block const& _block)
 
 void ContractCompiler::endVisit(Block const& _block)
 {
+	m_context.popVisitedNodes();
```
Comment (hrkrshnn) [inline libsolidity/codegen/ContractCompiler.cpp]: Where would the `popScopedVariables` 4 lines below point to now?

Comment (chriseth) [reply libsolidity/codegen/ContractCompiler.cpp]: I don't see a problem here. `popVisitedNodes` is mainly for source locations, and the AST will not change anyway.

Comment (hrkrshnn) [reply libsolidity/codegen/ContractCompiler.cpp]: I meant the source location. Shouldn't the source location of the pop-ed variables be in the block?

Comment (chriseth) [reply libsolidity/codegen/ContractCompiler.cpp]: Ah, now I understand, so this is actually not the real fix...

I'll move it down, thanks for noticing!

**Comment by chriseth** [comment]
After the fix, there is no change in commandline tests. I'll try to see if I can add a test for the specific issue.

**Comment by chriseth** [comment]
Added a test.

**Review by hrkrshnn — APPROVED** [review]

**Pull request merged**